lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -654,10 +660,191 @@ namespace Intl
             ret = 0;
         }
 
-LCloseCollator:
         ucol_close(collator);
         return ret;
     }
+
+    // Gets the system time zone. If tz is non-null, it is written into tz
+    // Returns the number of characters written into tz (including a null terminator)
+    int GetDefaultTimeZone(_Out_writes_opt_(tzLen) char16 *tz, _In_ int tzLen)
+    {
+        UErrorCode status = U_ZERO_ERROR;
+        int required = ucal_getDefaultTimeZone(reinterpret_cast<UChar *>(tz), tzLen, &status);
+        if (tz == nullptr && tzLen == 0 && status == U_BUFFER_OVERFLOW_ERROR)
+        {
+            // buffer overflow is expected when we are just trying to get the length returned
+            return required + 1;
+        }
+        
+        ICU_ASSERT(status, required > 0 && required < tzLen);
+        return required + 1;
+    }
+
+    // Determines if a time zone is valid. If it is and tzOut is non-null, the canonicalized version is written into tzOut
+    // Returns the number of characters written into tzOut (including a null terminator), or 0 on invalid time zone
+    int ValidateAndCanonicalizeTimeZone(_In_z_ const char16 *tzIn, _Out_writes_opt_(tzOutLen) char16 *tzOut, _In_ int tzOutLen)
+    {
+        UErrorCode status = U_ZERO_ERROR;
+        
+        // ucal_getCanonicalTimeZoneID does not like being passed nullptr/0-length buffers, even if we just want to get the expected buffer length
+        char16 tempBuf;
+        const int tempBufLen = 1;
+        bool hasTempBuf = false;
+        if (tzOut == nullptr && tzOutLen == 0)
+        {
+            tzOut = &tempBuf;
+            tzOutLen = tempBufLen;
+            hasTempBuf = true;
+        }
+
+        int required = ucal_getCanonicalTimeZoneID(reinterpret_cast<const UChar *>(tzIn), -1, reinterpret_cast<UChar *>(tzOut), tzOutLen, nullptr, &status);
+        if (hasTempBuf && status == U_BUFFER_OVERFLOW_ERROR)
+        {
+            // buffer overflow is expected when we are just trying to get the length returned
+            return required + 1;
+        }
+        else if (status == U_ILLEGAL_ARGUMENT_ERROR)
+        {
+            // illegal argument here means that tzIn is an invalid time zone
+            return 0;
+        }
+
+        ICU_ASSERT(status, tzOut == nullptr ? required > 0 : required <= tzOutLen);
```

Comment:
Making these asserts more clear in a future revision.

Reply:
How so? What's unclear? Could a comment help?

---
In reply to: [155606046](https://github.com/Microsoft/ChakraCore/pull/4390#discussion_r155606046) [](ancestors = 155606046)